core/payments/reservation/reservation_payment_validator.go

```diff
@@ -0,0 +1,85 @@
+package reservation
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/core/payments"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// Validates reservation payments for multiple accounts
+type ReservationPaymentValidator struct {
+	logger logging.Logger
+	// A cache of the ledgers being tracked
+	ledgerCache *ReservationLedgerCache
+	timeSource  func() time.Time
+}
+
+func NewReservationPaymentValidator(
+	ctx context.Context,
+	logger logging.Logger,
+	// The maximum number of ReservationLedger entries to be kept in the LRU cache. This may be automatically increased
+	// at runtime if premature ledger evictions are detected by the underlying cache.
+	maxLedgers int,
+	// provides access to payment vault contract
+	paymentVault payments.PaymentVault,
+	// source of current time for the leaky bucket algorithm
+	timeSource func() time.Time,
+	// how to handle requests that would overfill the bucket
+	overfillBehavior OverfillBehavior,
+	// duration used to calculate bucket capacity
+	bucketCapacityPeriod time.Duration,
+	// interval for checking for payment updates
+	updateInterval time.Duration,
+) (*ReservationPaymentValidator, error) {
+
+	ledgerCache, err := NewReservationLedgerCache(
+		ctx,
+		logger,
+		maxLedgers,
+		paymentVault,
+		timeSource,
+		overfillBehavior,
+		bucketCapacityPeriod,
+		updateInterval,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new reservation ledger cache: %w", err)
+	}
+
+	return &ReservationPaymentValidator{
+		logger:      logger,
+		ledgerCache: ledgerCache,
+		timeSource:  timeSource,
+	}, nil
+}
+
+// Validates a reservation payment for a blob dispersal
+// The caller is responsible for verifying the signature before calling this method
+func (pv *ReservationPaymentValidator) Debit(
+	ctx context.Context,
+	accountID gethcommon.Address,
+	symbolCount uint32,
+	quorumNumbers []uint8,
+	dispersalTime time.Time,
+) error {
+	ledger, err := pv.ledgerCache.GetOrCreate(ctx, accountID)
+	if err != nil {
+		return fmt.Errorf("get or create ledger: %w", err)
+	}
+
+	now := pv.timeSource()
+	success, _, err := ledger.Debit(now, dispersalTime, symbolCount, quorumNumbers)
```

Comment:
Bit of a race here, but you already called it out in a comment.

```
	// Don't bother checking if another routine already re-created this ledger. Even if another routine *did* create
	// a new instance, it's reasonable to preference the old instance over the new. There may be some small discrepancy
	// here, but there would be no feasible way for a malicious client to exploit this. In the worst case, the leaky
	// bucket will be slightly less filled than it ought to have been. Since it's incredibly unlikely to happen in the
	// first place, it's not worth contorting the design to address.
	c.cache.Add(accountID, reservationLedger)
```

Reply:
Do you think I need to add anything to this struct? Or is the doc in the cache enough?

Reply:
I'm ok with this in its current form